src/pe/tls.rs

```diff
@@ -223,8 +272,14 @@ impl<'a> TlsData<'a> {
                 if callback == 0 {
                     break;
                 }
+                if callback < image_base as u64 {
+                    return Err(error::Error::Malformed(format!(
+                        "tls callback ({:#x}) is less than image base ({:#x})",
+                        callback, image_base
+                    )));
+                }
                 // Each callback is an VA so convert it to RVA
-                let callback_rva = callback as usize - image_base;
+                let callback_rva = callback.wrapping_sub(image_base as u64) as usize;
```

Comment:
This is a bit strange but it seems like it requires it even we have checks above otherwise panics with subtract by overflow.

Probably compiler optimizations do something unpredicted. I may be able to reverse for further investigation.

Reply:
Well, that was my fault. I was not smart enough to consider `usize` v. `u64` behaves differently in the x86 binaries.

For this case, by using `wrapping_sub` it is casted correctly (64-bit var to 64-bit var), but without it, it casts to the 32-bit var and then 64-bit var, then subtract with overflow. Makes sense.

I'm going to fix this.

```rs
fn main() {
    println!("usize {}", core::mem::size_of::<usize>());
    println!("  u64 {}", core::mem::size_of::<u64>());
}
```

```bash
cargo run --target x86_64-pc-windows-msvc
usize 8
  u64 8
```

```bash
cargo run --target i686-pc-windows-msvc
usize 4
  u64 8
```

Reply:
Somewhat related, ‘image_base‘ in goblin should be ‘u64‘ instead of ‘usize‘ to be compatible with both x64 and x86 samples on each host builds.

That's implemented without aware of x86 host compatibility. Perhaps fixed in 0.10 but would likely to be subject to breaking change.